### PR TITLE
executor: fix a data race in dataReaderBuilder

### PR DESCRIPTION
**Pull request by tiancaiamao**
There is a data race in `dataReaderBuilder.buildExecutorForDatums()` when it call `builder()` :
`dataReaderBuilder.buildExecutorForDatums()` may be called in multiple goroutines, the share the common `*executorBuilder`.

When some goroutine is setting `executorBuilder.err` field during `build()`, other goroutine checking `executorBuilder.err`, data race will happen.

Fix it by extract common function buildIndexReader/buildTableReader/buildIndexLookUpReader out of builder object.

@coocood @zz-jason

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -975,16 +974,16 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plan.PhysicalIndexJoin) Execut
...
+	dagReq, err := constructDAGReq(v.TablePlans, b.getStartTS(), b.ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 	ts := v.TablePlans[0].(*plan.PhysicalTableScan)
 	table, _ := b.is.TableByID(ts.Table.ID)
-	newRanges := b.constructTableRanges(ts)
-	if b.err != nil {
-		return nil
+	newRanges, err := constructTableRanges(ts, b.ctx)
```
Comment (coocood) [inline executor/builder.go]: We can move the range construction out to `executorBuilder.buildTableReader`, we don't need range from the plan in `dataReaderBuilder`.

File: executor/builder.go [hunk 2]
```diff
@@ -1003,18 +1002,26 @@ func (b *executorBuilder) buildTableReader(v *plan.PhysicalTableReader) *TableRe
...
+func buildIndexReader(b *executorBuilder, v *plan.PhysicalIndexReader) (*IndexReaderExecutor, error) {
+	dagReq, err := constructDAGReq(v.IndexPlans, b.getStartTS(), b.ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 	is := v.IndexPlans[0].(*plan.PhysicalIndexScan)
-	newRanges := b.constructIndexRanges(is)
-	if b.err != nil {
-		return nil
+	newRanges, err := constructIndexRanges(is, b.ctx)
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 3]
```diff
@@ -1073,9 +1088,9 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
 		tableReq.OutputOffsets = append(tableReq.OutputOffsets, uint32(i))
 	}
 
-	newRanges := b.constructIndexRanges(is)
-	if b.err != nil {
-		return nil
+	newRanges, err := constructIndexRanges(is, b.ctx)
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 4]
```diff
@@ -1091,15 +1105,35 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
...
 		dataReaderBuilder: &dataReaderBuilder{executorBuilder: b},
 	}
-	return e
+	return e, nil
+}
+
+func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpReader) *IndexLookUpExecutor {
+	ret, err := buildIndexLookUpReader(b, v)
+	if err != nil {
+		b.err = err
```
Comment (zz-jason) [inline executor/builder.go]: `b.err = errors.Trace(err)`

File: executor/builder.go [hunk 5]
```diff
@@ -893,25 +893,24 @@ func (b *executorBuilder) buildAnalyze(v *plan.Analyze) Executor {
 	return e
 }
 
-func (b *executorBuilder) constructDAGReq(plans []plan.PhysicalPlan) *tipb.DAGRequest {
+func constructDAGReq(plans []plan.PhysicalPlan, startTS uint64, ctx context.Context) (*tipb.DAGRequest, error) {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/builder.go]: This context is not `goctx.Context`, but a SQL context (actually a session object), so it does not  violate Go convention.
we've rename all standard `context` to `goctx`.

File: executor/builder.go [hunk 6]
```diff
@@ -893,25 +893,24 @@ func (b *executorBuilder) buildAnalyze(v *plan.Analyze) Executor {
...
 		}
 		dagReq.Executors = append(dagReq.Executors, execPB)
 	}
-	return dagReq
+	return dagReq, nil
 }
 
-func (b *executorBuilder) constructTableRanges(ts *plan.PhysicalTableScan) (newRanges []types.IntColumnRange) {
-	sc := b.ctx.GetSessionVars().StmtCtx
+func constructTableRanges(ts *plan.PhysicalTableScan, ctx context.Context) (newRanges []types.IntColumnRange, err error) {
```
Comment (coocood) [inline executor/builder.go]: This is only used in executor builder, should be a method of it.

File: executor/builder.go [hunk 7]
```diff
@@ -922,29 +921,29 @@ func (b *executorBuilder) constructTableRanges(ts *plan.PhysicalTableScan) (newR
...
 		}
 		newRanges = ranger.Ranges2IntRanges(ranges)
 	}
-	return newRanges
+	return newRanges, nil
 }
 
-func (b *executorBuilder) constructIndexRanges(is *plan.PhysicalIndexScan) (newRanges []*types.IndexRange) {
-	sc := b.ctx.GetSessionVars().StmtCtx
+func constructIndexRanges(is *plan.PhysicalIndexScan, ctx context.Context) (newRanges []*types.IndexRange, err error) {
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 8]
```diff
@@ -975,17 +974,13 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plan.PhysicalIndexJoin) Execut
 	}
 }
 
-func (b *executorBuilder) buildTableReader(v *plan.PhysicalTableReader) *TableReaderExecutor {
-	dagReq := b.constructDAGReq(v.TablePlans)
-	if b.err != nil {
-		return nil
+func buildTableReader(b *executorBuilder, v *plan.PhysicalTableReader) (*TableReaderExecutor, error) {
```
Comment (coocood) [inline executor/builder.go]: buildNoRangeTableReader would be more clear.

File: executor/builder.go (hunk 6 above)
Comment (tiancaiamao) [reply executor/builder.go]: No, to be used, it must be a function, `buildTableReader` function need it.
If it become a method of executor builder, `builder.err` field will be modified and DATA RACE raise.

File: executor/builder.go [hunk 9]
```diff
@@ -1035,18 +1041,35 @@ func (b *executorBuilder) buildIndexReader(v *plan.PhysicalIndexReader) *IndexRe
...
+	newRanges, err1 := constructIndexRanges(is, b.ctx)
+	if err1 != nil {
+		b.err = errors.Trace(err1)
 		return nil
 	}
+	ret.ranges = newRanges
+	return ret
+}
+
+func buildIndexLookUpReader(b *executorBuilder, v *plan.PhysicalIndexLookUpReader) (*IndexLookUpExecutor, error) {
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 10]
```diff
@@ -1003,19 +997,32 @@ func (b *executorBuilder) buildTableReader(v *plan.PhysicalTableReader) *TableRe
...
+	newRanges, err1 := constructTableRanges(ts, b.ctx)
+	if err1 != nil {
+		b.err = errors.Trace(err1)
 		return nil
 	}
+	ret.ranges = newRanges
+	return ret
+}
+
+func buildIndexReader(b *executorBuilder, v *plan.PhysicalIndexReader) (*IndexReaderExecutor, error) {
```
Comment (coocood) [inline executor/builder.go]: ditto

File: executor/builder.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zz-jason) [reply executor/builder.go]: I still think it's better to move `context` to the first parameter, just to be compatible with other existed codes.

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/builder.go]: I'll replace it with `sc *variable.StatementContext`, so you'll not mind this any more...

File: executor/builder.go (hunk 6 above)
Comment (coocood) [reply executor/builder.go]: It doesn't have to be a function, we can still return the error instead of modifying builder.err.

**Comment by coocood** [comment]
LGTM

**Pull request merged**